File: src/functions/GetPostBySlug.ts
import { response } from '@libs/api-gateway';
import { PostService } from '../libs/post.service';
import { APIGatewayEvent, APIGatewayProxyHandler } from "aws-lambda";

export const handler: APIGatewayProxyHandler = async (
  event: APIGatewayEvent
): Promise<any> => {

  if (!event.pathParameters.slug) {
    throw new Error('slug is not defined')
  }

  const postService: PostService = new PostService();
  const post = await postService.getPostBySlug(event.pathParameters.slug);
  
  return response(200, post);
};

File: src/functions/GetPostBySlug.go
package functions

import (
	"context"
	"fmt"
	"net/http"

	"github.com/aws/aws-lambda-go/events"

	"blog-cms-v2/src/libs"
)

func GetPostBySlug(ctx context.Context, event events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
	slug := event.PathParameters["slug"]
	if slug == "" {
		return libs.ErrResponse(http.StatusBadRequest, fmt.Errorf("slug is not defined")), nil
	}

	postService := libs.NewPostService()
	post, err := postService.GetPostBySlug(ctx, slug)
	if err != nil {
		return libs.ErrResponse(http.StatusInternalServerError, fmt.Errorf("fetching post: %w", err)), nil
	}
	if post == nil {
		return libs.ErrResponse(http.StatusNotFound, fmt.Errorf("post not found: %s", slug)), nil
	}
	return libs.Response(http.StatusOK, post), nil
}
